pypicloud/cache/dynamo.py

```diff
@@ -22,6 +25,78 @@
 LOG = logging.getLogger(__name__)
 
 
+def _path_depth(package):
+    """Count the slashes in path, or return infinity on no path."""
+    path = package.data.get('path')
+    if not path:
+        return float('+inf')
+    return path.count('/')
+
+
+def _decide_between_versions(contender, current):
+    """Decide between packages with the same filename on different paths."""
+
+    if contender.last_modified < current.last_modified:
+        return contender
```

Comment:
Okay great, I'm glad we're on the same page. Line-by-line:

```
if cache.filename != storage.filename:
```
Pretty sure that should never happen, as we're only comparing packages with the same filename. Is harmless, though.

```
if cache.data['path'] == storage.data['path']:
    return contender
```

√
(but use .get() because 'path' doesn't exist for all storage backends)

```
  return sorted(
    [cache, storage],
    key=lambda o: (
      o.last_modified,
      o.data['path'],
    ),
  )[0]
```

Not sure exactly the reasoning behind this. I think that there are two possible goals for the case where the paths are different:

1) **Least surprise to the package consumer**: In this case we should always `return current` because anything else will change the `path` stored in the cache, and pypicloud will begin serving a *different* file from S3. Preferably spit out a log line indicating what has happened.

2) **Same data as "absolute reset"**: In this case we should always `return contender` because anything else would leave different data in the cache & storage.

There is a separate but related issue that you pointed out, which is how to decide between multiple files from storage that have the same filename. Right now I don't think there's a code path that handles that at all. Since we're keeping `updated_packages` in a set, the set behavior is what's going to decide. The package `__eq__` method compares package name and version (though maybe it should be changed to compare filename) so I think the winner will be the first file returned by S3.

All of the caches currently have undefined behavior in this case, so I am okay with leaving it as such. If you want to make it deterministic then I'm fine with sorting on basically any field.